### PR TITLE
Final Exam Submission

### PR DESCRIPTION
**Pull request by NuclearWoofer**

### CODE REVIEW

File: lopezMidterm/src/components/Navigation.tsx [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
 import { NavLink } from 'react-router-dom';
 
-function Navigation() {
+const Navigation: React.FC = () => {
   return (
     <nav>
       <ul style={{ listStyle: 'none', display: 'flex', gap: '10px', color: '' }}>
         <li>
-          <NavLink exact to="/" activeclassname="active">
+          <NavLink to="/" activeClassName="active">
```
Comment (bowen9284) [inline lopezMidterm/src/components/Navigation.tsx:9]: `activeClassName` is from an older version of React Router and doesn't work on version 6. The IDE tells you it's an invalid property as well. Use caution when using other resources and when in doubt look at the official documentation.

File: lopezMidterm/src/types/cartTypes.ts
Comment (bowen9284) [inline lopezMidterm/src/types/cartTypes.ts:1]: Nice 👍🏻

File: lopezMidterm/src/reducers/cartReducer.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+interface CartState {
+  items: CartItem[];
+}
+
+const initialState: CartState = {
+  items: [],
+};
+
+const cartReducer = (state = initialState, action: CartAction): CartState => {
```
Comment (bowen9284) [inline lopezMidterm/src/reducers/cartReducer.ts:11]: You have a little bit of a mix of Redux-toolkit (RTK) and the older version of Redux. Some of the features will work together so this is okay. In RTK, we don't have to worry about keeping our state immutable because it's done behind the scenes for us.

File: lopezMidterm/src/reducers/cartReducer.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { ADD_TO_CART, REMOVE_FROM_CART, CLEAR_CART, CartAction, CartItem } from '../types/cartTypes';
+
+interface CartState {
+  items: CartItem[];
+}
+
+const initialState: CartState = {
+  items: [],
```
Comment (bowen9284) [inline lopezMidterm/src/reducers/cartReducer.ts:8]: You could use `items.length` to display the number of items in your cart in your header. The main reason that Redux exists is for global state. By showing the number of cart items, we are demonstrating that we can share state across multiple screens.

File: lopezMidterm/src/components/Checkout.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+const Checkout: React.FC = () => {
+  const cartItems = useSelector((state: RootState) => state.cart.items);
+  const dispatch = useDispatch();
+  const [totalPrice, setTotalPrice] = useState(0);
+
+  useEffect(() => {
+    const fetchPrices = async () => {
+      try {
+        const response = await axios.get('/products');
```
Comment (bowen9284) [inline lopezMidterm/src/components/Checkout.tsx:25]: Should be using a custom hook here instead of axios.get directly in the component.

File: lopezMidterm/src/components/Checkout.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      } catch (error) {
+        console.error('Error fetching prices:', error);
+      }
+    };
+
+    fetchPrices();
+  }, [cartItems]);
+
+  useEffect(() => {
+    const updatedTotal = cartItems.reduce(
```
Comment (bowen9284) [inline lopezMidterm/src/components/Checkout.tsx:44]: Since we can use Redux, this is a perfect candidate for deriving state using a selector. you can learn more about that  [here](https://redux.js.org/usage/deriving-data-selectors#calculating-derived-data-with-selectors).

TLDR; a reusable function that is tied to your global state.

File: lopezMidterm/src/components/ProductDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  const handleAddToCart = () => {
+    if (product) {
+      dispatch(addToCart(product));
+      window.alert(`${product.title} added to cart`);
+    }
+  };
+
+  return (
+    <>
+      {loading ? (
```
Comment (bowen9284) [inline lopezMidterm/src/components/ProductDetails.tsx:33]: Return early like we have been doing in other exercises. Nested ternaries are tough to read and are prone to errors.

```
if (loading) return <Loading />
if (error) return <Error />

if (product) {
  // product JSX
} else {
  <p>No product found</p>
}
```

File: lopezMidterm/src/hooks/useFetch.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import { useEffect, useState } from 'react';
+import axios, { AxiosResponse } from 'axios';
+import React from 'react';
+
+interface FetchResult<T> {
+  data: T;
```
Comment (bowen9284) [inline lopezMidterm/src/hooks/useFetch.tsx:6]: Make sure you understand what `T` is, how it differs from the `any` type and why you would use it here instead of any`

File: lopezMidterm/src/main.jsx
Comment (bowen9284) [inline lopezMidterm/src/main.jsx:1]: This should be `main.tsx`

File: package.json [hunk 8]
```diff
@@ -9,6 +9,7 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
-    "axios": "^1.4.0"
+    "axios": "^1.4.0",
+    "prop-types": "^15.8.1"
```
Comment (bowen9284) [inline package.json:13]: You're missing the Typescript dependency

File: lopezMidterm/src/components/Products.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import './Products.css';
+
+interface Product {
+  id: number;
+  title: string;
+  image: string;
+}
+
+const Products: React.FC = () => {
+  const { data: products, loading, error } = useFetch<Product[]>('/products');
```
Comment (bowen9284) [inline lopezMidterm/src/components/Products.tsx:13]: if you hover over `products` it will be of type `Product[][]` which causes a lot of TS errors down below. To fix this just drop the array brackets after useFetch `useFetch<Product>`. That's because you told useFetch it's already returning an array.

This would be cleaner as it's own custom hook `useFetchProduct` and we'd return data as type products.

File: lopezMidterm/src/components/Products.tsx [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  const [visibleProducts, setVisibleProducts] = useState<number>(8);
+
+  const handleShowMore = () => {
```
Comment (bowen9284) [inline lopezMidterm/src/components/Products.tsx:16]: I like this idea of a sort of pagination 👍🏻

File: lopezMidterm/src/hooks/useContext.tsx [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { useContext } from 'react';
+import { ThemeContext } from '../context/ThemeContext';
+
+export const useTheme = () => {
+  const { theme, toggleTheme } = useContext(ThemeContext);
+  return { theme, toggleTheme };
```
Comment (bowen9284) [inline lopezMidterm/src/hooks/useContext.tsx:6]:
```suggestion
export const useTheme = () => {
  return useContext(ThemeContext);
```

File: lopezMidterm/src/hooks/useContext.tsx
Comment (bowen9284) [inline lopezMidterm/src/hooks/useContext.tsx:1]: Always try to match the file name with the name of the component and/or hook

**Review by bowen9284 — APPROVED** [review]
Thanks Michael. You did a good job with the redux portion, although I feel like you could have went a little further with it. I left some suggestions in the pull request. The two biggest things that hurt you were missing a theme and missing some important pieces of Typescript. 

You made it clear that it was not implemented, but I did see most of the code there so I gave you a few points instead of 0.

The Typescript dependency was not installed and our configuration files are missing. Visual Studio Code and the Vite project can give you a false sense of security. The idea here was to go back to what we learned when we converted our app and apply it here.

See rubric for more details about your grade.

78/100
